mrmustard/lab/representations/data/matvec_data.py

```diff
@@ -0,0 +1,183 @@
+# Copyright 2023 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+import numpy as np
+from numba import njit
+from typing import List
+from mrmustard.lab import Data
+from mrmustard.math import Math
+from mrmustard.typing import Batch, Matrix, Scalar, Vector
+
+math = Math()
+
+
+class MatVecData(Data):  # Note : this class is abstract too!
+    def __init__(
+        self, mat: Batch[Matrix], vec: Batch[Vector], coeffs: Batch[Scalar]
+    ) -> None:
+        self.mat = math.atleast_3d(mat)
+        self.vec = math.atleast_2d(vec)
+        self.coeffs = math.atleast_1d(coeffs)
+
+    @property
+    def batch_size(self) -> int:
+        return self.coeffs.shape[0]
+
+
+    #@njit
+    def __neg__(self) -> MatVecData:
+        return self.__class__(mat=-self.mat, vec=-self.vec, coeffs=self.coeffs)
+
+
+    def __eq__(self, other: MatVecData, rtol: float = 1e-6, atol: float = 1e-6) -> bool:
+
+        try: 
+            return super().same(
+                X = [self.mat, self.vec, self.coeffs],
+                Y = [other.mat, other.vec, other.coeffs],
+                rtol = rtol,
+                atol = atol
+                ) 
+
+        except AttributeError as e:
+            raise TypeError(
+                f"Cannot compare {self.__class__} and {other.__class__}.") from e
+
+
+    #@njit(parallel=True)
+    def __add__(
+        self,
+        other: MatVecData,
+        rtol: float = 1e-6,
+        atol: float = 1e-6,
+        check_for_equality: bool = False,
+        sub:bool = False # TODO : find a more elegant way to subtract! cf below
```

Comment:
`__add__` only takes self and other as arguments. In particular, don't implement `sub` this way, just add self and -other.